### PR TITLE
mds: reject sessionless messages

### PR DESCRIPTION
**Pull request by XiaoGuoDong2019**
mds should reject sessionless messages.

Fixes: https://tracker.ceph.com/issues/41329
Signed-off-by: Xiao Guodong xiaogd@inspur.com

### CODE REVIEW

**Comment by tchaikov** [comment]
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/Server.cc: In member function 'void Server::handle_client_session(ceph::cref_t<MClientSession>&)':
/home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/Server.cc:445:76: error: 'MClientSession::MClientSession(int, version_t)' is protected within this context
     m->get_connection()->send_message(new MClientSession(CEPH_SESSION_CLOSE));
                                                                            ^
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/Server.h:25,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/MDSRank.h:37,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/Server.cc:22:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/messages/MClientSession.h:40:3: note: declared protected here
   MClientSession(int o, version_t s=0) :
   ^~~~~~~~~~~~~~
src/mds/CMakeFiles/mds.dir/build.make:153: recipe for target 'src/mds/CMakeFiles/mds.dir/Server.cc.o' failed
make[3]: *** [src/mds/CMakeFiles/mds.dir/Server.cc.o] Error 1
```

@XiaoGuoDong2019 ^

**Comment by liwei021111** [comment]
> ```
> /home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/Server.cc: In member function 'void Server::handle_client_session(ceph::cref_t<MClientSession>&)':
> /home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/Server.cc:445:76: error: 'MClientSession::MClientSession(int, version_t)' is protected within this context
>      m->get_connection()->send_message(new MClientSession(CEPH_SESSION_CLOSE));
>                                                                             ^
> In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/Server.h:25,
>                  from /home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/MDSRank.h:37,
>                  from /home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/Server.cc:22:
> /home/jenkins-build/build/workspace/ceph-pull-requests/src/messages/MClientSession.h:40:3: note: declared protected here
>    MClientSession(int o, version_t s=0) :
>    ^~~~~~~~~~~~~~
> src/mds/CMakeFiles/mds.dir/build.make:153: recipe for target 'src/mds/CMakeFiles/mds.dir/Server.cc.o' failed
> make[3]: *** [src/mds/CMakeFiles/mds.dir/Server.cc.o] Error 1
> ```
> 
> @XiaoGuoDong2019 ^

please try make again.

**Comment by XiaoGuoDong2019** [comment]
> ```
> /home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/Server.cc: In member function 'void Server::handle_client_session(ceph::cref_t<MClientSession>&)':
> /home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/Server.cc:445:76: error: 'MClientSession::MClientSession(int, version_t)' is protected within this context
>      m->get_connection()->send_message(new MClientSession(CEPH_SESSION_CLOSE));
>                                                                             ^
> In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/Server.h:25,
>                  from /home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/MDSRank.h:37,
>                  from /home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/Server.cc:22:
> /home/jenkins-build/build/workspace/ceph-pull-requests/src/messages/MClientSession.h:40:3: note: declared protected here
>    MClientSession(int o, version_t s=0) :
>    ^~~~~~~~~~~~~~
> src/mds/CMakeFiles/mds.dir/build.make:153: recipe for target 'src/mds/CMakeFiles/mds.dir/Server.cc.o' failed
> make[3]: *** [src/mds/CMakeFiles/mds.dir/Server.cc.o] Error 1
> ```
> 
> @XiaoGuoDong2019 ^

please make again, thanks

File: src/mds/Server.cc [hunk 1]
```diff
@@ -442,6 +442,8 @@ void Server::handle_client_session(const cref_t<MClientSession> &m)
 
   if (!session) {
     dout(0) << " ignoring sessionless msg " << *m << dendl;
+    auto reply = make_message<MClientSession>(CEPH_SESSION_CLOSE);
+    mds->send_message(reply, m->get_connection());
```
Comment (batrick) [inline src/mds/Server.cc]: Better would be to send a reject message, see the code for handling blacklisted clients below in this method.

Comment (batrick) [reply src/mds/Server.cc]: You can also add the metadata for whwy the session is rejected. See:

`m->metadata["error_string"] = "sessionless";`
 
in the below code.

**Comment by XiaoGuoDong2019** [comment]
@batrick please review, thanks.

**Pull request merged**